pkg/kubeclient/client.go

```diff
@@ -0,0 +1,89 @@
+package kubeclient
+
+import (
+	"context"
+
+	"k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/watch"
+)
+
+// Interface defines capability of the Kubernetes Client.
+type Interface interface {
+	// Create creates given object in Kubernetes cluster
+	Create(ctx context.Context, obj runtime.Object) error
+
+	// Get retrieves given object from Kubernetes cluster
+	Get(ctx context.Context, obj runtime.Object) error
+
+	// Update updates given object in Kubernetes cluster
+	Update(ctx context.Context, obj runtime.Object) error
+
+	// Delete deletes given object from Kubernetes cluster
+	Delete(ctx context.Context, obj runtime.Object, opts *metav1.DeleteOptions) error
+
+	// List retrieves list of object
+	List(ctx context.Context, obj runtime.Object, opts *metav1.ListOptions) error
+
+	// TODO: define capabilities for ListDeletion, Patch and Watch ?
+}
+
+// NewFromConfigOrCluster returns a Kubernetes client. If configPath is
+// provided, it uses the configPath to initialize a client else it assumes to be
+// running within a Kubernetes cluster and tries to initialize with cluster
+// configuration.
+func FromConfigOrCluster(configPath string) (Interface, error) {
```

Comment:
Is `FromConfigOrCluster` going to call `FromConfig` and `IsCluster`?

Reply:
`Cluster` is confusing if you mean `InCluster`. Also the documentation reads `NewFrom...` but the method name is `From...`. TBH, I like the `NewFrom` form better. (also applies to the function below).

Reply:
> and tries to initialize with **in**-cluster configuration

Reply:
> If configPath is provided, it uses the configPath to initialize a client else it assumes

Why would I call this method if I don't provide a `configPath`, rather than calling the `InCluster` directly?

Reply:
Now we have only one constructor which takes rest.Config as input. There are already helper methods in client-go which lets you create rest.Config.